include/swift/ABI/TaskStatus.h

```diff
@@ -252,16 +252,71 @@ class CancellationNotificationStatusRecord : public TaskStatusRecord {
   using FunctionType = SWIFT_CC(swift) void (SWIFT_CONTEXT void *);
 
 private:
+  static const uintptr_t StatusMask = 0x2;
+
   FunctionType * __ptrauth_swift_cancellation_notification_function Function;
-  void *Argument;
+  std::atomic<uintptr_t> Storage; // Argument and Status
+
+  /// Status used to ensure that we only execute the cancellation handler *once*
+  ///
+  /// The race that this is intended to resolve is the following:
+  /// as the task itself is adding the cancellation record it performs:
+  /// - add (this) cancellation record
+  /// - check again, if we were cancelled immediately *after* adding the task
+  ///   the cancelling task could not have picked up the cancellation record
+  ///   and therefore would have not executed it yet. If we indeed were
+  ///   cancelled there, we must execute the handler. However, we cannot be
+  ///   certain if the cancelling task did, or did not see the cancellation
+  ///   record - since it is executing concurrently. To resolve this race,
+  ///   running the handler flips the Pending bit to Executed, and therefore
+  ///   regardless if the cancelling task did run the handler, or if we must
+  ///   do it we will never double-execute a handler.
+  ///
+  ///   Skipping the second isCancelled check in the executing task is not
+  ///   safe, because it could lead to adding the record, then becoming
+  ///   cancelled and therefore never running the cancellation handler at
+  ///   all.
+  struct ExecutedStatus {
+    static const uintptr_t Pending = 0b0;
+    static const uintptr_t Executed = 0b1;
+  };
+
+  struct ArgumentAndStatus {
+    uintptr_t Storage;
+
+    uintptr_t getStatus() const { return Storage & StatusMask; }
+
+    bool wasExecuted() const { return Storage & ExecutedStatus::Executed; }
+    bool canExecute() const { return !wasExecuted(); }
+
+    void *getArgument() const {
+      return reinterpret_cast<void *>(Storage & ~StatusMask);
```

Comment:
I'm not quite sure how this works. Is the pointer guaranteed to never have the LSB set?

Reply:
Yeah, we do that in many places in the task infra. 

Turns out we won't need this dance, but yeah this would have been right bitwise